### ISSUE TITLE
[Feature Request] Ability to pin files or directories to the top

### ISSUE THREAD

**Issue by aksr** [vifm/vifm#1008]
One thing I often had a need for is an **ability to pin some files or directories to the top**.

Since filters (:filter, =, ...) are fairly inexpensive, I have hundreds (thousands etc.) of files of different type etc. in one (e.g. home) directory, and, usually, *no* need (at least not for a long time and depending on the work I do at the moment) to hierarchically structure them.

Sometimes, after many days or much time spent on a specific thing, some *other* thing cames up to which I need to pay attention, it gets lost *down the files* (I have ":set sort=-ctime,dir" by default)

I would propose, if it's *feasible* and *doesn't require* too much *complex coding*, 
to add an ability to pin files or directories (e.g. [dD]ownloads) to the top.

Every new pinned file could go to the top of the list of pinned files
(pinned files would be immune to the sorting by default, we could have this settable by option (e.g. :set sortpinned..)).
If a pinned file is *again pinned*, it could go to the top of the pinned files.

We could, maybe latter, generalize this further...

**Comment by xaizek** [comment]
The previous time it came up I suggested to [use 'sortgroups' option](https://q2a.vifm.info/1514/pin-folder-to-top?show=1515#a1515) as a workaround, but it has its limitations.

Providing this as a builtin functionality requires some more analysis.  I'm not sure I fully understand how pinned items would relate to sorting.  At the moment it sounds like they would just ignore it completely (including reversing of sort order), so it wouldn't be a modification of `name` sorting key in `'sort'` but a modification of processing `'sort'` overall (this makes a difference for when you don't sort by name).  I guess in tree-view pinned items still respect tree structure but bubble up within corresponding level, and similarly for ls-like view.

Also, items are to be pinned within a specific path, right?

**Comment by xaizek** [comment]
Thanks, things are clearer now.

> Maybe we could have _two kinds of pinning_: one _within a specific path_ and the other _across all directories_.

If path matching for pinning is done via [patterns](https://vifm.info/vimdoc.shtml#vifm-patterns), then there is no distinction because patterns can match either file names or paths.  However, that might be too much for pinning where something more specialized could suffice.

> Also, rules could be defined for the second one in **vifmrc**.

If there is some kind of `:pin` command, it might as well do any kind of pinning regardless of the used syntax.

Order of pinned files can probably match relative order of corresponding patterns/paths so that reordering them would result in different file order.

**Comment by xaizek** [comment]
> `:pin soft` would be influenced by sorting etc.

In other words, not pinned?  Is that supposed to unpin things?

**Comment by xaizek** [comment]
> `:pin soft` would be movable (meaning effected by sort).

If the purpose of pinning is for entries to not be affected by sorting, then how come this is pinning?

**Comment by xaizek** [comment]
> Have you started on a prototype and/or made some kind of progress to implement this?

No, don't expect it in the next release.  It's already rather late and I still have long list of things to do for it.
